keps/sig-api-machinery/20190802-dynamic-coordinated-storage-version.md

```diff
@@ -172,6 +172,35 @@ During bootstrap, for each resource, the API server
   to avoid conflicting with other API servers.
 * installs the resource handler.
 
+The above mentioned process requires an API server to update the storageVersion
+before accepting API requests. If we don't enforce this order, data encoded in
+an unexpected version can sneak into etcd. For example, an API server persists a
+write request encoded in an obsoleted version, then it crashes before it can
+update the storageVersion. The storage migrator has no way to detect this write.
+
+For the cmd/kube-apiserver binary, we plan to enforce this order by adding a new
+filter to the [handler chain][]. Before kube-aggregator, kube-apiserver, and
+apiextension-apiserver have registered the storage version of the built-in
+resources they host, this filter only allows the following requests to pass:
+1. a request sent by the loopbackClient and is destined to the storageVersion
+   API.
+2. the verb of the request is GET.
+3. the request is for an API that is not persisted, e.g.,
+   SubjectAccessReview and TokenReview. [Here] is a complete list.
+4. the request is for an aggregated API or an API defined by CRD.
```

Comment:
I think you have to block CRD requests since the kube-apiserver is the one doing that write.

Reply:
> I think you have to block CRD requests since the kube-apiserver is the one doing that write.

Update around "needs to update the storageVersion object when it [creates][] the custom" to indicate that you need to write the storageversion.status *before* adding the CR handler so that the ordering is correct.

Please split the aggregated API and CRD numbers here and reference the section for why we can serve CRDs.  It's safe because we ensure the storageversion is correct before we serve it, same as the stock APIs we have.